builder/builder.go

```diff
@@ -209,17 +209,54 @@ func (b *Builder) Build(ctx context.Context, payload *Payload) (*monomer.Block,
 	if err := b.txStore.Add(txResults); err != nil {
 		return nil, fmt.Errorf("add tx results: %v", err)
 	}
+
 	// Publish events.
+	if err := b.publishEvents(txResults, block, resp); err != nil {
+		return nil, fmt.Errorf("publish events: %v", err)
+	}
+
+	return block, nil
+}
+
+func (b *Builder) publishEvents(txResults []*abcitypes.TxResult, block *monomer.Block, resp *abcitypes.ResponseFinalizeBlock) error {
+	numTxs := block.Txs.Len()
+	// Initial capacity for the block events slice is conservatively set to the number of txs in the block. This will
+	// dynamically expand as necessary as block events are added to it.
```

Comment:
Yeah I thought about just starting with an empty slice, however that would require a lot more reallocations since every tx will have at least 1-2 events.

Reply:
After re-reading through the [docs](https://docs.cosmos.network/main/learn/advanced/events#default-events) it looks like every cosmos message is guaranteed to have at least 3 default events, in addition to any events added by modules. Every cosmos tx can contain one or more messages.

After looking through this again, I think a conservative initial capacity for the block events slice would be the number of txs * 3 and then have the slice dynamically reallocate as needed. Lmk if you guys agree or feel differently.